lib/moonshot/command_line.rb

```diff
@@ -0,0 +1,108 @@
+require 'thor'
+
+module Moonshot
+  # This class implements the command-line `moonshot` tool.
+  class CommandLine
+    # TODO: No reason this is class-level, make this a for-real class that needs to be initialized.
+    def self.run!
+      # Find the Moonfile in this project.
+      orig_dir = Dir.pwd
+
+      loop do
+        break if File.exist?('Moonfile')
+
+        if Dir.pwd == '/'
+          warn 'Could not find Moonfile!'
+          exit 1
+        else
+          Dir.chdir('..')
+        end
+      end
+
+      moonfile_dir = Dir.pwd
+      Dir.chdir(orig_dir)
+
+      # Load any plugins and CLI extensions relative to the Moonfile
+      if File.directory?(File.join(moonfile_dir, 'moonshot'))
+        load_cli_extensions(moonfile_dir)
+        load_plugins(moonfile_dir)
+      end
+
+      Object.include(Moonshot::ArtifactRepository)
+      Object.include(Moonshot::BuildMechanism)
+      Object.include(Moonshot::DeploymentMechanism)
+      load(File.join(moonfile_dir, 'Moonfile'))
+
+      load_commands
+
+      # Determine what command is being run, which should be the first argument.
+      command = ARGV.shift
+      usage if %w[--help -h help].include?(command) or command.nil?
+
+      # Dispatch to that command, by executing it's parser, then
+      # comparing ARGV to the execute methods arity.
+      unless @commands.key?(command)
+        warn "Command not found: #{command}"
+        warn "\n"
+        usage
+      end
+
+      handler = @commands[command].new
+      handler.parser.parse!
+
+      unless ARGV.size == handler.method(:execute).arity
+        warn handler.parser.help
+        exit 1
+      end
+
+      handler.execute(*ARGV)
+    end
+
+    def self.register(klass)
+      @classes ||= []
+      @classes << klass
+    end
+
+    def self.registered_commands
+      @classes || []
+    end
+
+    def self.load_plugins(moonfile_dir)
+      plugins_path = File.join(moonfile_dir, 'moonshot', 'plugins', '**', '*.rb')
+      Dir.glob(plugins_path).each do |file|
+        load(file)
+      end
+    end
+
+    def self.load_cli_extensions(moonfile_dir)
+      cli_extensions_path = File.join(moonfile_dir, 'moonshot', 'cli_extensions', '**', '*.rb')
+      Dir.glob(cli_extensions_path).each do |file|
+        load(file)
+      end
+    end
+
+    def self.usage
+      warn "Usage: moonshot [command]"
+      warn
+      warn "Valid commands include:"
+      @commands.each do |c, k|
+        warn "#{c} #{k.description}"
+      end
```

Comment:
This looked a little dense compared to other tools, some whitespace might be nice.

Global options would be nice too.

```
$ cargo --help
Rust's package manager

Usage:
    cargo <command> [<args>...]
    cargo [options]

Options:
    -h, --help          Display this message
    -V, --version       Print version info and exit
    --list              List installed commands
    --explain CODE      Run `rustc --explain CODE`
    -v, --verbose ...   Use verbose output
    -q, --quiet         No output printed to stdout
    --color WHEN        Coloring: auto, always, never

Some common cargo commands are (see all commands with --list):
    build       Compile the current project
    clean       Remove the target directory
    doc         Build this project's and its dependencies' documentation
    new         Create a new cargo project
    init        Create a new cargo project in an existing directory
    run         Build and execute src/main.rs
    test        Run the tests
    bench       Run the benchmarks
    update      Update dependencies listed in Cargo.lock
    search      Search registry for crates
    publish     Package and upload this project to the registry
    install     Install a Rust binary

See 'cargo help <command>' for more information on a specific command.

$ composer --help
Usage:
 help [--xml] [--format="..."] [--raw] [command_name]

Arguments:
 command               The command to execute
 command_name          The command name (default: "help")

Options:
 --xml                 To output help as XML
 --format              To output help in other formats (default: "txt")
 --raw                 To output raw command help
 --help (-h)           Display this help message
 --quiet (-q)          Do not output any message
 --verbose (-v|vv|vvv) Increase the verbosity of messages: 1 for normal output, 2 for more verbose output and 3 for debug
 --version (-V)        Display this application version
 --ansi                Force ANSI output
 --no-ansi             Disable ANSI output
 --no-interaction (-n) Do not ask any interactive question
 --profile             Display timing and memory usage information
 --working-dir (-d)    If specified, use the given directory as working directory.

Help:
 The help command displays help for a given command:
   composer help list

 You can also output the help in other formats by using the --format option:

   composer help --format=xml list

 To display the list of available commands, please use the list command.
```